app/Jobs/CreateQueryserviceBatchesJob.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace App\Jobs;
+
+use App\EventPageUpdate;
+use App\QsBatch;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Config;
+
+class CreateQueryserviceBatchesJob extends Job
+{
+    private const NAMESPACE_ITEM = 120;
+    private const NAMESPACE_PROPERTY = 122;
+    private const NAMESPACE_LEXEME = 146;
+
+    private int $entityLimit;
+
+    public function __construct()
+    {
+        $this->entityLimit = Config::get('wbstack.qs_batch_entity_limit');
+    }
+
+    public function handle(): void
+    {
+        DB::transaction(function () {
+            // Get ID of the latest batch that we have created (or 0)
+            $maxIdBatch = QsBatch::orderBy('id', 'desc')->first();
+            if ($maxIdBatch) {
+                $batchesUpToEventId = $maxIdBatch->eventTo;
+            } else {
+                // If there are no batches then things just hav not really started yet
+                $batchesUpToEventId = 0;
+            }
+
+            // Get events after the point that batch was created
+            // TODO maybe filter by NS here?
+            $events = EventPageUpdate::where('id', '>', $batchesUpToEventId)->get();
+
+            $wikiBatchesEntities = [];
+            $lastEventId = 0;
+            foreach ($events as $event) {
+                if (
+                    $event->namespace == self::NAMESPACE_ITEM ||
+                    $event->namespace == self::NAMESPACE_PROPERTY ||
+                    $event->namespace == self::NAMESPACE_LEXEME
+                ) {
+                    $wikiBatchesEntities[$event->wiki_id][] = $event->title;
+                }
+                if ($event->id > $lastEventId) {
+                    $lastEventId = $event->id;
+                }
+            }
+
+            $notDoneBatches = QsBatch::where([
+                ['done', '=', 0], ['pending_since', '=', null], ['failed', '=', false]
+            ])->get();
+
+            // Insert the newly created batches into the table...
+            foreach ($wikiBatchesEntities as $wikiId => $entityBatch) {
+                // If we already have a not done batch for this same wiki, then merge that into a new batch
+                foreach ($notDoneBatches as $qsBatch) {
+                    if ($qsBatch->wiki_id == $wikiId) {
+                        $entitiesOnBatch = explode(',', $qsBatch->entityIds);
+                        if (count($entitiesOnBatch) >= $this->entityLimit) {
+                            continue;
+                        }
+                        $entityBatch = array_merge($entityBatch, $entitiesOnBatch);
+                        // Delete the old batch
+                        $qsBatch->delete();
```

Comment:
I wonder if this is what we want. As we delete the batch and create a new one here, it will affect the sorting done in the QsController (which happens by id).

Alternatively, we could update the batch in such a case and keep the existing ordering.

Reply:
> Perhaps a solution is to follow up and sort in the controller by oldest edit time on the QSBatch model (and not on id)

Not sure if I understand all the details here (sadly after such a long time working on this), but considering:

- the id is an incrementing int
- we never update models, but will delete and create a new one on update

wouldn't sorting on `created_at` and `id` give us the exact same results?

Reply:
> if we update the batch we need to rejig the "cursor" logic you mentioned in the resolved comment above I think. 

In my understanding, it's the other way around. If we update, we get a "stable" ordering of wikis in the queue, if we delete and create we have wikis being pushed back in the queue each time they see an update.

Reply:
This test case might help visualizing what's currently shuffling things around: https://github.com/wbstack/api/blob/a58d8a03b7075a4e232900e97ffa251d80bd3181/tests/Jobs/CreateQueryserviceBatchesJobTest.php#L123